stdlib/domain.ml

```diff
@@ -112,3 +112,35 @@ let get_id { domain; _ } = domain
 
 let self () = Raw.self ()
 
+module TLS = struct
+
+  type 'a key = int ref
+
+  external get_primitive : unit -> (int ref * Obj.t) list
+    = "caml_domain_tls_get"
+
+  external set_primitive : (int ref * Obj.t) list  -> unit
+    = "caml_domain_tls_set"
+
+  let new_key () = let v = ref 0 in v
+
+  let set k x =
+    let cs = Obj.repr x in
+    let old = get_primitive () in
+    set_primitive @@ (k,cs)::old
```

Comment:
Multiple updates to the same key will create multiple entries in the list. The list will keep growing. Instead of 

```
(int ref * Obj.t) list
```

use 

```
(int ref * Obj.t ref) list
```

`set` should first check that the key is already present in the list. If so, update the second element of the pair with the new value. Otherwise, push to the front of the list. It may be nicer to do:

```
type entry = { key : int ref; slot : Obj.t ref }

external get_primitive : unit -> entry list
    = "caml_domain_tls_get"

  external set_primitive : entry list  -> unit
    = "caml_domain_tls_set"
```

A value of type `entry` has the same memory representation as `(int ref * Obj.t ref)`.

Reply:
Growing list is indeed going to be a problem. But updating a reference inside a list was getting quite tricky. I have now removed old associated values (if any present) and then updated the new entry, so the list now has only one value associated with a key.

Another possible solution to this could be using a hash table that looks something like this:
```ocaml
module RefHash =
    struct
      type t = int ref
      let equal i j = i == j
      let hash = (* need a suitable hash function here, default hash function will end 
      up with many collisions since we are using `ref 0` as all the keys*) 
    end
```
As long as there aren't too many variables in the thread-local state, I think the current implementation is okay.

Reply:
I've left a comment. The current implementation using `remove_assq` is inefficient. I'd still recommend going to use the ref.